feature: Support transforming named column(s) in any table

carlos-granados opened the pull request:
Adds a new type of transformer that is able to transform a column in any table. For example you might have this scenario:

```
scenario: Check user age
  Then the users ages should be:
    | user   |  age  |
    | john   |  21   |
    | peter  |  34   |
```
And this transformer:
``` php
#[Transforrm('column:user')]
public function transformUsernameToUser(string $username): User
{
    return $this->getUserByUsername($username);
}
```
Then this transformer will be called to transform any data in the column named `user`

The final result which is passed to the step that uses this transformer is an array of rows where the data for the column `user` will be the transformed user. So, for example the step could be implemented like this:

```php
#[Then('the users ages should be:')]
public function checkUsersAges(array $table): void
{
    foreach ($table as $row) {
        Assert::assertSame($row['user']->getAge(), $row['age']);
    }
} 
```

This transformation can be applied to different tables as long as the column name matches.

The transformation can be applied to more than one column if you list more than one column name in the Transform definition, for example:
``` php
#[Transforrm('column:user,other user')]
```

More than one transformation can be applied to different columns of a table, so in the example above if we had another transformation defined for the `age` column it would also be applied. The end result is again an array where in each row any column that matched any of the transformations has been replaced with its transformed value

Before implementing this new feature I modified the existing tests for the transformations to adopt the new format where the tests are defined using real files in the fixtures folder. This has been done in a separate previous commit to make it easier to review

With this feature we can get the functionality described in https://github.com/Behat/Behat/issues/1229

Once this is merged, I will update the documentation for this transformer (and also other transformers like the row transformer which are not documented)

carlos-granados commented on tests/Fixtures/Transformations/behat.php:
```diff
@@ -0,0 +1,158 @@
+
+return (new Config())
+    ->withProfile((new Profile('attributes'))
+        ->withSuite((new Suite('simple_step_argument_transformation'))
```
This file defines the features and contexts that we use for each test

acoulton commented on features/definitions_transformations_annotations.feature:
```diff
@@ -4,160 +4,17 @@ Feature: Step Arguments Transformations Annotations
+      | option      | value       |
+      | --no-colors |             |
+      | --format    | progress    |
+      | --profile   | annotations |
```
nice solution!

acoulton commented on tests/Fixtures/Transformations/features/bootstrap/UserAnnotationsContext.php:
```diff
@@ -0,0 +1,80 @@
+use Behat\Behat\Context\Context;
+use PHPUnit\Framework\Assert;
+
+class UserAnnotationsContext implements Context
```
I wonder if this is strictly necessary? We have separate feature coverage that annotations work for steps, these features "officially" only need to test the annotations and attributes for transforms. So it'd probably be fine to just use the `UserAttributesContext` for the **step** definitions in both profiles?

acoulton commented on features/definitions_transformations_annotations.feature:
```diff
@@ -29,10 +29,10 @@ Feature: Step Arguments Transformations Annotations
-      3 scenarios (3 passed)
-      9 steps (9 passed)
+      4 scenarios (4 passed)
+      12 steps (12 passed)
```
Might be just Monday morning, but I'm not following why the scenario count has changed for this feature and the row table argument transformations feature in the second commit?

carlos-granados replied:
I rebased the code to include the changes I had added for the transformers with unicode characters and I included those changes in the existing commits but I made a weird split, I added the changes to the files in the fixtures in the first commit but only added the changes to the main feature file in the second commit 😬 . Hope this makes sense, if not I can try to rebase this again and split it better

acoulton replied:
Makes sense, I think it's fine, just wanted to make sure there wasn't some unexpected change / I wasn't missing something.

acoulton commented on src/Behat/Behat/Transformation/Transformation/TableColumnTransformation.php:
```diff
@@ -0,0 +1,136 @@
+        $argumentIndex,
+        $argumentArgumentValue
+    ): bool {
+        if (!$argumentArgumentValue instanceof TableNode && !is_array($argumentArgumentValue)) {
```
Might be useful as a code comment?

acoulton commented on tests/Fixtures/Transformations/features/bootstrap/TransformationAnnotationsContext.php:
```diff
@@ -79,6 +79,26 @@ public function createUserNamesFromTableInRussian($tableRow) {
         return $tableRow['логин'];
     }
 
+    /** @Transform column:user,other user */
+    public function convertUsernameToUserInColumn(string $name): User {
+        return new User($name);
+    }
+
+    /** @Transform column:логин */
+    public function convertUsernameInRussianToUserInColumn(string $name): User {
+        return new User($name);
+    }
+
+    /** @Transform column:username */
+    public function convertUsernameToUserWithUsernameHeading(string $name): User {
+        throw new Exception('This should not be called');
+    }
+
+    /** @Transform column:hex age */
+    public function convertHexAgeToAge(string $hexAge): int {
+        return hexdec($hexAge);
+    }
+
```
It might be better to define these in the attributes feature / profile instead of the annotations one? Just means fewer things to move if/when we drop annotations, and since it's a new feature it's probably reasonable to test only with attributes?

acoulton commented on src/Behat/Behat/Transformation/Transformation/TableColumnTransformation.php:
```diff
@@ -0,0 +1,136 @@
+
+        if ($argumentArgumentValue instanceof TableNode) {
+            $tableHeadings = $argumentArgumentValue->getRow(0);
+            foreach ($columnNames as $columnName) {
+                if (in_array($columnName, $tableHeadings, true)) {
+                    return true;
+                }
+                return false;
+            }
```
May be more efficient and make the method slightly easier to follow?

```suggestion
            return array_intersect($columnNames, $tableHeadings) !== [];            
```

acoulton commented on src/Behat/Behat/Transformation/Transformation/TableColumnTransformation.php:
```diff
@@ -0,0 +1,136 @@
+                return false;
+            }
+        }
+        foreach ($argumentArgumentValue as $row) {
```
If the argument started as a `TableNode`, can we ever have a case where a row in the array has different keys to the row in `$argumentArgumentValue[0]`?

carlos-granados replied:
Probably not, but added this check in case some other transformer interferes with the results

acoulton commented on src/Behat/Behat/Transformation/Transformation/TableColumnTransformation.php:
```diff
@@ -0,0 +1,136 @@
+     */
+    public function getPriority()
+    {
+        return 30;
```
Could be a useful code comment

acoulton reviewed (commented):
Another super-helpful feature, thanks @carlos-granados! Just a few suggestions / comments from me :)

carlos-granados commented:
@acoulton I rebased this code and added two commits:
- One with coding standard changes detected by our new linter
- Another with changes following your suggestions

acoulton commented:
@carlos-granados I think maybe the commit with your followup to code review suggestions hasn't been pushed? As there are a few things you've :+1: but I don't see changed in the diff. Thanks :)

carlos-granados commented:
> @carlos-granados I think maybe the commit with your followup to code review suggestions hasn't been pushed? As there are a few things you've 👍 but I don't see changed in the diff. Thanks :)

Argh, yes forgot to push, done!

acoulton reviewed (approved):
Great work, thanks @carlos-granados :)
